url0e7d57473a_kyorohiro_HetimaUtil_tgz-pJ8-net_hetimatan_net_http_HttpServer3xxJ8/src/net/hetimatan/util/http/HttpResponse.java

```diff
@@ -0,0 +1,237 @@
+		return mStatusCode;
+	}
+
+	public String getReasonPharse() {
+		return mReasonPharse;
+	}
+
+	public KyoroFile getContent() {
+		return mContent;
+	}
+
+
+	public void setContent(byte[] content) throws IOException {
+		setContent(new CashKyoroFile(content));
+	}
+
+	public void setContent(KyoroFile content) {
+		mContent = content;
+	}
+
+	public LinkedList<HttpRequestHeader> getHeader() {
+		return mHeaders;
+	}
+
+	public String getHeader(String key) {
+		for(HttpRequestHeader header : mHeaders) {
+			if(key.replaceAll(" ", "").toLowerCase().equals(header.getKey().replaceAll(" ", "").toLowerCase())){
+				return header.getValue();
+			}
+		}
+		return "";
+	}
+	public HttpResponse addHeader(String key, String value) {
+		return addHeader(new HttpRequestHeader(key, value));
+	}
+
+	public HttpResponse addHeader(HttpRequestHeader header) {
+//		System.out.println("key="+header.getKey()+","+header.getValue());
+		if (header.getKey().toLowerCase()
+				.equals(HttpRequestHeader.HEADER_CONTENT_LENGTH.toLowerCase())) {
+			mContentLength = header;
+		}
+		mHeaders.add(header);
+		return this;
+	}
+
+	public long getContentSizeFromHeader() {
+		if(mContentLength == null) {
+			return Integer.MAX_VALUE;
+		} 
+		String value = mContentLength.getValue();
+		try {
+			return Integer.parseInt(value.replaceAll(" ", ""));
+		} catch(Exception e) {
+			return 0;
+		}
+	}
+
+	@Override
+	public void encode(OutputStream output) throws IOException {
+		output.write((mHttpVersion + SP + mStatusCode + SP + mReasonPharse + CRLF)
+				.getBytes());
+		long contentLength = 0;
+		if(mContent != null) {
+			contentLength = mContent.length();
+		}
+		output.write(("" + HttpRequestHeader.HEADER_CONTENT_LENGTH + ": "
+				+ contentLength + CRLF).getBytes());
+		for (HttpRequestHeader header : mHeaders) {
+			header.encode(output);
+		}
+		output.write(CRLF.getBytes());
+
+		byte[] buffer = new byte[100];
+		while (contentLength!=0) {
+			int len = mContent.read(buffer);
+			if (len < 0) {
+				break;
+			}
+			output.write(buffer, 0, len);
+		}
+		output.flush();
+	}
+
+	// Status-Line = HTTP-Version SP Status-Code SP Reason-Phrase CRLF
+	public static HttpResponse decode(MarkableReader reader, boolean isReadBody) throws IOException {
+		String httpVersion = "";
+		String statusCode = "";
+		String reasonPhrase = "";
+
+		httpVersion = _httpVersion(reader);
+		_sp(reader);
+		statusCode = _statusCode(reader);
+		_sp(reader);
+		reasonPhrase = _reasonPhrase(reader);
+		_crlf(reader);
+		HttpResponse ret = new HttpResponse(httpVersion, statusCode, reasonPhrase);
+
+		
+		try {
+			while (true) {
+				if (isCrlf(reader)) {
+					break;
+				}
+				ret.addHeader(HttpRequestHeader.decode(reader));
+			}
+		} catch (IOException e) {
+		}
+
+		_crlf(reader);
+
+		long size = ret.getContentSizeFromHeader();
+		if(size <0) {size = Integer.MAX_VALUE;}
+		// 
+		// todo MarkableReader内の VFを使用してもよいかも
+		CashKyoroFile vf = new CashKyoroFile(512, 2);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
Analysis of this package determined that this line of code might contain a resource leak. A resource leak can slow down or crash your system.

**Fix**
Consider closing the following resource in a try-finally block: *vf*. Currently, there are execution paths that do not contain closure statements, for example, when MarkableReader.read throws an exception.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).